src/precice/tests/SerialTests.cpp

```diff
@@ -547,6 +547,69 @@ BOOST_AUTO_TEST_CASE(testExplicitWithDataInitialization)
   }
 }
 
+/**
+ * @brief Tests the reading and writing of data multiple times within one timestep.
+ *
+ * The first solver performs multiple consistent readings of data sent by the second solver.
+ * The second solver performs multiple sendings, of which the last is expected by the first solver.
+ */
+BOOST_AUTO_TEST_CASE(testExplicitWithDataMultipleReadWrite)
+{
+  PRECICE_TEST("SolverOne"_on(1_rank), "SolverTwo"_on(1_rank));
+
+  using Eigen::Vector3d;
+
+  SolverInterface cplInterface(context.name, _pathToTests + "explicit-data-multiple-read-write.xml", 0, 1);
+  if (context.isNamed("SolverOne")) {
+    int meshOneID = cplInterface.getMeshID("MeshOne");
+    cplInterface.setMeshVertex(meshOneID, Vector3d(1.0, 1.0, 1.0).data());
+    double maxDt     = cplInterface.initialize();
+    int    dataID    = cplInterface.getDataID("DataOne", meshOneID);
+    double valueData = 0.0;
+    cplInterface.initializeData();
+    cplInterface.readScalarData(dataID, 0, valueData);
+    // expected data value received
+    BOOST_TEST(3.0 == valueData);
+    // change value at read destination
+    valueData = -0.2;
+    BOOST_TEST(-0.2 == valueData);
+    cplInterface.readScalarData(dataID, 0, valueData);
+    // expected data value received
+    BOOST_TEST(3.0 == valueData);
+    while (cplInterface.isCouplingOngoing()) {
+      maxDt = cplInterface.advance(maxDt);
+      cplInterface.readScalarData(dataID, 0, valueData);
+      // expected data value received
+      BOOST_TEST(5.0 == valueData);
+      // change value at read destination
+      valueData = -0.4;
+      BOOST_TEST(-0.4 == valueData);
+      cplInterface.readScalarData(dataID, 0, valueData);
+      // expected data value received
+      BOOST_TEST(5.0 == valueData);
+    }
+    cplInterface.finalize();
+  } else {
+    BOOST_TEST(context.isNamed("SolverTwo"));
+    int meshTwoID = cplInterface.getMeshID("MeshTwo");
+    cplInterface.setMeshVertex(meshTwoID, Vector3d(0.0, 0.0, 0.0).data());
```

Comment:
Here, you should use the return value of `setMeshVertex`. Below the ID `0` just pops up, but actually you should get this ID from `setMeshVertex`.